### ISSUE TITLE
Failed in get: Hummock error: ObjectStore failed with IO error 

### ISSUE THREAD

**Issue by lmatz** [risingwavelabs/risingwave#7002]
### Describe the bug

Slack link:
https://risingwave-labs.slack.com/archives/C048NM5LNKX/p1671603133726859
Namespace: `rwc-3-longevity-20221220-180642`
Pod: `risingwave-compute-2`

```
2022-12-20T18:11:34.875003Z ERROR risingwave_storage::monitor::monitored_store: Failed in get: Hummock error: ObjectStore failed with IO error Internal error: read "rls-apse1-eks-a-rwc-3-longevity-20221220-180642/255/1.data" in block Some(BlockLocation { offset: 11008083, size: 37158 }) failed, error: timeout: error trying to connect: HTTP connect timeout occurred after 3.1s
  backtrace of `ObjectError`:
   0: <risingwave_object_store::object::error::ObjectError as core::convert::From<risingwave_object_store::object::error::ObjectErrorInner>>::from
             at ./risingwave/src/object_store/src/object/error.rs:38:10
   1: <T as core::convert::Into<U>>::into
             at ./rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/core/src/convert/mod.rs:726:9
   2: <risingwave_object_store::object::error::ObjectError as core::convert::From<aws_smithy_http::result::SdkError<E>>>::from
             at ./risingwave/src/object_store/src/object/error.rs:81:9
   3: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at ./rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/core/src/result.rs:2108:27
   4: <risingwave_object_store::object::s3::S3ObjectStore as risingwave_object_store::object::ObjectStore>::read::{{closure}}
             at ./risingwave/src/object_store/src/object/s3.rs:351:20
   5: <async_stack_trace::StackTraced<F,_> as core::future::future::Future>::poll
   6: risingwave_object_store::object::MonitoredObjectStore<OS>::read::{{closure}}
             at ./risingwave/src/object_store/src/object/mod.rs:643:13
   7: risingwave_object_store::object::ObjectStoreImpl::read::{{closure}}
             at ./risingwave/src/object_store/src/object/mod.rs:334:9
   8: risingwave_storage::hummock::sstable_store::SstableStore::sstable::{{closure}}::{{closure}}::{{closure}}
             at ./risingwave/src/storage/src/hummock/sstable_store.rs:346:25
   9: risingwave_common::cache::LruCache<K,T>::lookup_with_request_dedup::{{closure}}::{{closure}}
             at ./risingwave/src/common/src/cache.rs:818:58
  10: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at ./root/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
```

### Additional context
Or this is expected?

**Comment by lmatz** [comment]
happened again in https://risingwave-labs.slack.com/archives/C048NM5LNKX/p1672121457286999

Namespace: `rwc-3-longevity-20221226-180525`
Pod: `risingwave-compute-1`

**Comment by zwang28** [comment]
Some info:
- We are using 8C16G node in this test. Though we cannot tell their EC2 instance types because of lost test env, I think they are very likely with a baseline bandwidth of 2.5 Gbps and a burst bandwidth of 10 Gbps, according to those 2xlarge instance types.
- When error occurs, network IO bytes do appear close to 2.5 Gbps baseline bandwidth, as shown below.
  Out bytes
  <img width="1447" alt="out" src="https://user-images.githubusercontent.com/70626450/210042789-bfecb47e-586a-4d6e-9cf7-6bd57033cba7.png">
  In bytes
  <img width="1446" alt="in" src="https://user-images.githubusercontent.com/70626450/210042793-86b6adee-3df6-434d-8c8a-e80e113c5091.png">

Let's wait and see another test running with higher network capacity.

**Comment by liurenjie1024** [comment]
Why we don't wait and retry rather than panicking here?

**Comment by zwang28** [comment]
> Why we don't wait and retry rather than panicking here?

It's retried 3 times [here](https://github.com/risingwavelabs/risingwave/blob/44b4c10100a47b2add0199e87e710092e6aeb4ee/src/object_store/src/object/s3.rs#L520), which doesn't work out.

**Comment by zwang28** [comment]
In rwc-3-longevity-20230104-180851 we are using c5a.8xlarge (10 Gbps network capacity) for compute nodes.
The rate of (node_network_transmit_bytes + node_network_receive_bytes) does reach 10Gbps.
We'd better impose other strategy besides merely increasing node's network capacity.

**Comment by liurenjie1024** [comment]
> In rwc-3-longevity-20230104-180851 we are using c5a.8xlarge (10 Gbps network capacity) for compute nodes. The rate of (node_network_transmit_bytes + node_network_receive_bytes) does reach 10Gbps. We'd better impose more strategy besides merely increasing node's network capacity.

+1. We should add random latency to retry rather than panicking here.

**Comment by zwang28** [comment]
Increase the [connect_timeout](https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/aws-smithy-types/src/timeout.rs#L30) (3.1s by default) does work around this issue. 
But I'm not sure it's a good idea to increase it. Can this kind of IO error be more of an indicator that current cluster size cannot handle the workload, and we should consider to reduce the load of current worker nodes?

**Comment by lmatz** [comment]
>But I'm not sure it's a good idea to increase it. 

Agree not a good idea

> Can this kind of IO error be more of an indicator that current cluster size cannot handle the workload, and we should consider to reduce the load of current worker nodes?

As the kernel manages CPU resources by setting parallelism not more than the number of CPUs, and manages memory resources by having `GlobalMemoryManager` evict states from time to time, both in a proactive way,

it feels sort of strange if it manages network resources in a reactive way, although being proactive seems a more difficult task indeed.

made-up cases:
1. it's a temporary increase in the usage of the network, and probably we just slow down the processing for a short period of time and then we can get back to normal again, e.g. another form of backpressure. Rescheduling or scaling in/out may cost too much when it comes to this kind of network usage fluctuation, or it's not allowed because (2)
or
2. users only want to allocate a certain of resources for some jobs but are willing to tolerate lower throughput and higher latency

**Comment by fuyufjh** [comment]
Seems to be caused by insufficient bandwidth

**Comment by zwang28** [comment]
rwc-3-longevity-20230131-171156
seems not caused by bandwidth this time.

**Comment by zwang28** [comment]
Lots of `WARN risingwave_storage::hummock::iterator::merge_inner: PeekMut are dropped without used. May be caused by future cancellation` before this IO error.

[compute-1.csv](https://github.com/risingwavelabs/risingwave/files/10554947/compute-1.csv)

**Comment by zwang28** [comment]
In terms of network bandwidth, neither of these 2 failed case reaches 10Gbps limit.

rwc-3-longevity-20230131-171156
<img width="1442" alt="rwc-3-longevity-20230131-171156" src="https://user-images.githubusercontent.com/70626450/216289510-c23a535f-c551-482f-8902-462a16b981c1.png">

rwc-3-longevity-20230201-170952
<img width="1456" alt="rwc-3-longevity-20230201-170952" src="https://user-images.githubusercontent.com/70626450/216290083-4aceeb1f-5a5c-4a11-a239-ae1a9b593604.png">

**Comment by zwang28** [comment]
- [Each EC2 instance can send 1024 packets per second per network interface to Route 53 Resolver](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-dns.html#vpc-dns-limits)
- S3 SDK uses hyper, which uses [std's to_socket_addrs](https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html#tymethod.to_socket_addrs) to resolve address.
- S3 is an external DNS resources can be answered by [DNS cache](https://aws.amazon.com/premiumsupport/knowledge-center/dns-resolution-failures-ec2-linux/) 
- TODO have we enabled DNS cache?

**Comment by zwang28** [comment]
We encountered another case, where network transmit bytes/transmit packet number/current established TCP are all small.
The error keeps occurs, until we restart the compute nodes and the error is gone immediately 🤔 . The restarted computes nodes are still in the same k8s nodes.

**Comment by lmatz** [comment]
#8796

**Comment by zwang28** [comment]
fixed by #9160
